src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Collect.cs

```diff
@@ -0,0 +1,492 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Functions.DLP;
+using Microsoft.PowerFx.Core.Functions.FunctionArgValidators;
+using Microsoft.PowerFx.Core.IR;
+using Microsoft.PowerFx.Core.IR.Nodes;
+using Microsoft.PowerFx.Core.IR.Symbols;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+using RecordNode = Microsoft.PowerFx.Core.IR.Nodes.RecordNode;
+
+namespace Microsoft.PowerFx.Core.Texl.Builtins
+{
+    // Collect(collection:*[...], item1:![...]|*[...], ...)
+    internal class CollectFunction : BuiltinFunction, ISuggestionAwareFunction
+    {
+        public override bool AffectsCollectionSchemas => true;
+
+        public bool CanSuggestThisItem => true;
+
+        public override bool SupportsParamCoercion => false;
+
+        public override bool ManipulatesCollections => true;
+
+        public override bool ModifiesValues => true;
+
+        public override bool IsSelfContained => false;
+
+        public override bool RequiresDataSourceScope => true;
+
+        public virtual bool IsScalar => false;
+
+        public override bool CanSuggestInputColumns => true;
+
+        public override bool MutatesArg(int argIndex, TexlNode arg) => argIndex == 0;
+
+        public override RequiredDataSourcePermissions FunctionPermission => RequiredDataSourcePermissions.Create;
+
+        public override bool IsRestrictedUDFName => true;
+
+        /// <summary>
+        /// Since Arg1 and Arg2 depends on type of Arg1 return false for them.
+        /// </summary>
+        public override bool TryGetTypeForArgSuggestionAt(int argIndex, out DType type)
+        {
+            if (argIndex == 0)
+            {
+                type = DType.EmptyTable;
+                return true;
+            }
+            else if (argIndex == 1 || argIndex == 2)
+            {
+                type = default;
+                return false;
+            }
+            else
+            {
+                return base.TryGetTypeForArgSuggestionAt(argIndex, out type);
+            }            
+        }
+
+        public override bool ArgMatchesDatasourceType(int argNum)
+        {
+            return argNum >= 1;
+        }
+
+        public override bool IsLazyEvalParam(TexlNode node, int index, Features features)
+        {
+            // First argument to mutation functions is Lazy for datasources that are copy-on-write.
+            // If there are any side effects in the arguments, we want those to have taken place before we make the copy.
+            return index == 0;
```

Comment:
Does this change codegen for canvas?

Reply:
It didn't affect any test cases in Canvas. Can you think of a specific test in Canvas so we can check this?